db/write_batch.cc

```diff
@@ -1292,13 +1190,18 @@ Status WriteBatch::PopSavePoint() {
   return Status::OK();
 }
 
-Status WriteBatch::AssignTimestamp(const Slice& ts) {
-  TimestampAssigner ts_assigner(ts, prot_info_.get());
+Status WriteBatch::AssignTimestamp(
+    const Slice& ts, std::function<Status(uint32_t, size_t&)>&& checker) {
+  TimestampAssigner<decltype(checker)> ts_assigner(prot_info_.get(),
```

Comment:
We could avoid the overhead of `std::function` by turning `AssignTimestamp` / `AssignTimestamps` themselves into function templates that take the type `Checker` as a template parameter. (Right now, `TimestampAssigner` and `SimpleListTimestampAssigner` are only instantiated with a single type, namely the rvalue reference `std::function<Status(uint32_t, size_t&)>&&`.)

Reply:
Done.